extensions/basic_auth/plugin.cc

```diff
@@ -0,0 +1,330 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "extensions/basic_auth/plugin.h"
+
+#include <google/protobuf/io/zero_copy_stream_impl_lite.h>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+#include "extensions/common/json_util.h"
+
+#ifndef NULL_PLUGIN
+
+#include "extensions/metadata_exchange/base64.h"
+
+#else
+
+#include "common/common/base64.h"
+
+namespace proxy_wasm {
+namespace null_plugin {
+namespace BasicAuth {
+namespace Plugin {
+
+PROXY_WASM_NULL_PLUGIN_REGISTRY;
+
+using Base64 = Envoy::Base64;
+
+#endif
+
+using ::nlohmann::json;
+using ::Wasm::Common::JsonArrayIterate;
+using ::Wasm::Common::JsonGetField;
+using ::Wasm::Common::JsonValueAs;
+
+static RegisterContextFactory register_BasicAuth(
+    CONTEXT_FACTORY(PluginContext), ROOT_FACTORY(PluginRootContext));
+
+namespace {
+void deniedNoAccessToCredentials() {
+  sendLocalResponse(401,
+                    "Request denied by Basic Auth check. No credential "
+                    "has accessed to specified prefix path.",
+                    "", {});
+}
+
+void deniedNoAuthorizationHeader() {
+  sendLocalResponse(401,
+                    "Request denied by Basic Auth check. No Authorization "
+                    "header was found.",
+                    "", {});
+}
+
+void deniedNoBasicAuthData() {
+  sendLocalResponse(401,
+                    "Request denied by Basic Auth check. No Basic "
+                    "Authentication information found.",
+                    "", {});
+}
+
+void deniedInvalidCredentials() {
+  sendLocalResponse(401,
+                    "Request denied by Basic Auth check. Invalid "
+                    "username and/or password",
+                    "", {});
+}
+
+void deniedNoAcessToMethod() {
+  sendLocalResponse(401,
+                    "Request denied by Basic Auth check. No method "
+                    "has access to specified prefix path.",
```

Comment:
Consider rephrasing: Request method not authorized.

Reply:
sure!